File: problems/chapter16/miiitomi/001.cpp
// https://atcoder.jp/contests/abc010/tasks/abc010_4

#include <bits/stdc++.h>
using namespace std;

struct Edge {
    int rev, to, cap;
    Edge(int r, int t, int c) {
        rev = r;
        to = t;
        cap = c;
    }
};

bool rec(vector<vector<Edge>> &G, vector<bool> &seen, int v) {
    if (seen[v]) return false;
    seen[v] = true;
    if (v == (int)G.size()-1) return true;

    for (auto &e : G[v]) {
        if (e.cap == 0) continue;
        if (rec(G, seen, e.to)) {
            G[e.to][e.rev].cap++;
            e.cap--;
            return true;
        }
    }
    return false;
}

int FordFulkerson(vector<vector<Edge>> &G) {
    int ans = 0;
    while (true) {
        vector<bool> seen((int)G.size(), false);
        if (rec(G, seen, 0)) ans++;
        else break;
    }
    return ans;
}

int main() {
    int N, g, e;
    cin >> N >> g >> e;
    N++;

    vector<vector<Edge>> G(N);
    for (int i = 0; i < g; i++) {
        int p;
        cin >> p;
        G[p].push_back(Edge((int)G[N-1].size(), N-1, 1));
        G[N-1].push_back(Edge((int)G[p].size()-1, p, 1));
    }

    for (int i = 0; i < e; i++) {
        int a, b;
        cin >> a >> b;
        G[a].push_back(Edge((int)G[b].size(), b, 1));
        G[b].push_back(Edge((int)G[a].size()-1, a, 1));
    }

    cout << FordFulkerson(G) << endl;
}

File: problems/chapter16/mshibatatt/001.py
# https://atcoder.jp/contests/abc010/tasks/abc010_4
import sys
sys.setrecursionlimit = 10000

def dfs(G, now, g, path, seen):
    for i, nxt in enumerate(G[now]):
        if seen[nxt['to']] or nxt['weight'] == 0:
            continue
        if nxt['to'] == g:
            path.append([now, i])
            return path
        seen[nxt['to']] = True
        path.append([now, i])
        ret_path = dfs(G, nxt['to'], g, path, seen)
        if ret_path is not None:
            return ret_path

def add_edge(G, v_from, v_to, w):
    content =  {'to': v_to,
                'rev': len(G[v_to]),
                'weight': w
                }
    content_rev =  {'to': v_from,
                    'rev': len(G[v_from]),
                    'weight': 0
                    }
    G[v_from].append(content)
    G[v_to].append(content_rev)

def get_residual(G, path, flow):
    for i, j in path:
        G[i][j]['weight'] -= flow
        G[G[i][j]['to']][G[i][j]['rev']]['weight'] += flow
    return G

def main():
    N, G, E = map(int, input().split())
    girls = list(map(int, input().split()))
    SNS = [[] for _ in range(N+1)]
    for _ in range(E):
        a, b = map(int, input().split())
        add_edge(SNS, a, b, 1)
        add_edge(SNS, b, a, 1)
    for i in girls:
        add_edge(SNS, i, N, 1)
    
    # calculate edge-connectivity
    output = 0
    SNS_residual = SNS.copy()
    while True:
        seen = [False]*(2*N)
        seen[0] = True
        path = dfs(SNS_residual, 0, N, [], seen)
        if path is None:
            break
        SNS_residual = get_residual(SNS_residual, path, 1)
        output += 1
    
    print(output)

if __name__=='__main__':
    main()